plonky2/src/recursion/cyclic_recursion.rs

```diff
@@ -0,0 +1,490 @@
+#![allow(clippy::int_plus_one)] // Makes more sense for some inequalities below.
+use anyhow::{ensure, Result};
+use itertools::Itertools;
+use plonky2_field::extension::Extendable;
+
+use crate::gates::noop::NoopGate;
+use crate::hash::hash_types::{HashOut, HashOutTarget, MerkleCapTarget, RichField};
+use crate::hash::merkle_tree::MerkleCap;
+use crate::iop::target::{BoolTarget, Target};
+use crate::iop::witness::{PartialWitness, Witness};
+use crate::plonk::circuit_builder::CircuitBuilder;
+use crate::plonk::circuit_data::{
+    CircuitData, CommonCircuitData, VerifierCircuitTarget, VerifierOnlyCircuitData,
+};
+use crate::plonk::config::Hasher;
+use crate::plonk::config::{AlgebraicHasher, GenericConfig};
+use crate::plonk::proof::{ProofWithPublicInputs, ProofWithPublicInputsTarget};
+use crate::recursion::conditional_recursive_verifier::dummy_proof;
+
+pub struct CyclicRecursionData<
+    'a,
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+> {
+    proof: &'a Option<ProofWithPublicInputs<F, C, D>>,
+    verifier_data: &'a VerifierOnlyCircuitData<C, D>,
+    common_data: &'a CommonCircuitData<F, C, D>,
+}
+
+pub struct CyclicRecursionTarget<const D: usize> {
+    pub proof: ProofWithPublicInputsTarget<D>,
+    pub verifier_data: VerifierCircuitTarget,
+    pub dummy_proof: ProofWithPublicInputsTarget<D>,
+    pub dummy_verifier_data: VerifierCircuitTarget,
```

Comment:
Can these be baked into the circuit? Seems like they could be computed without running into a circular dependency.

Reply:
Hm what do you mean by "baked into the circuit"? Like as constants? What would be the advantage of that?

Reply:
Right I mean as constants; was thinking it could be nice if we didn't need to keep track of these targets and populate them later.

I probably should have suggested this for `conditionally_verify_proof_or_dummy` instead. There if we baked in `dummy_proof`, `dummy_verifier_data` as constants, we wouldn't need to return anything. So the API would match `verify_proof` better, and the user wouldn't need to call `set_proof_with_pis_target` or anything to populate dummy proof data.

With `cyclic_recursion` there'd be less advantage; the user would still need to call `set_cyclic_recursion_data_target` since it does other stuff besides the dummy proof.

I suppose the current approach is a bit more efficient currently, since it avoids extra `ConstantGate`s. Though it could change if we added custom select gates later.

Anyway definitely not something we need to change now.

Reply:
I see. I think adding all these constants could non-trivially increase the circuit size. Another option could be to non-deterministically set these targets using generators. That way we don't have to manually populate the targets and we also don't need constants.

Reply:
Good idea, adding generators seems like a nice solution for it

Reply:
I'll implement that in a later PR.